pkg/apis/core/types.go

```diff
@@ -3184,22 +3184,33 @@ type ClaimSource struct {
 	//
 	// The template will be used to create a new ResourceClaim, which will
 	// be bound to this pod. When this pod is deleted, the ResourceClaim
-	// will also be deleted. The name of the ResourceClaim will be <pod
-	// name>-<resource name>, where <resource name> is the
-	// PodResourceClaim.Name. Pod validation will reject the pod if the
-	// concatenated name is not valid for a ResourceClaim (e.g. too long).
-	//
-	// An existing ResourceClaim with that name that is not owned by the
-	// pod will not be used for the pod to avoid using an unrelated
-	// resource by mistake. Scheduling and pod startup are then blocked
-	// until the unrelated ResourceClaim is removed.
+	// will also be deleted. The ResourceClaim will have a name that is
+	// generated with <pod name>-<resource name> as base and the Pod as
```

Comment:
You probably want to spec here what happens if the `<pod name>-<resource name>` template is too long to fit in a claim name.

OK to do in a followup

Reply:
The concatenated string is stored in `claim.GeneratedName` and then the normal truncation happens when generating the actual name. I started to get worried that validation of `claim.GeneratedName` might reject a too long string, but that doesn't seem to be the case. I added an E2E test with `pod name = strings.Repeat("p", 63)` and `resource name =  strings.Repeat("c", 63)` and that worked:
```
    generateName: ppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppp-ccc
cccccccccccccccccccccccccccccccccccccccccccccccccccccccccccc
    name: pppppppppppppppppppppppppppppppppppppppppppppppppppppppppptcgdb
```

I'll add a comment about that.

Reply:
```
	// The template will be used to create a new ResourceClaim, which will
	// be bound to this pod. When this pod is deleted, the ResourceClaim
	// will also be deleted. The ResourceClaim will have <pod
	// name>-<resource name> as `generateName` base and the Pod as
	// owner. The final name then gets generated by the apiserver (i.e.
	// unique random suffix at the end, if necessary after truncating the
	// base). It gets recorded in pod.status.resourceClaimStatuses.
```

Reply:
ok.  The only quesion I have is: GenerateName has very high probability of succeeding at making something random, but in the case of long input, it just truncates.  That's USUALLY acceptable, but the alternative would be for you to do your own truncation, like:

```
if len(podName) + len(resName) > 57 { // 57 + "-<5random>"
    // truncate podName and resName in some more helpful way
}
```

Probably not worth it now, but may consider later.

Reply:
If we put "will have <pod name>-<resource name> as `generateName` base" into the API description, can we still change it later without an API break? Clients might have started to rely on exactly that aspect (for whatever reason). Should the exact way of truncating perhaps be left out of the API description after all or be marked as "may change"?

I like doing something better, for example truncating both strings separately. Whatever it is, clients should not rely on the exact process.

Reply:
Maybe just spec it as "The pod name and resource name, along with a generated component, will be used to form a unique name for the ResourceClaim, which will be recorded in..."

Also, from a safety point of view, consider what happens if the ResourceClaim is created, and then the controller crashes before recording the resulting name.  Does the ResourceClaim have a label sufficient to re-assemble the state (e.g. a k8s-owned key with the pod's UID as a value)?

Reply:
I like that wording better than my current one - changed.

Regarding safety: there is an annotation that allows the controller to continue. See [prior discussion with Jordan](https://github.com/kubernetes/kubernetes/pull/117351#discussion_r1253101683).